1-Draft/RFCNNNN-Improve-Background-Operator.md

```diff
@@ -0,0 +1,208 @@
+
+The `&` background operator has great potential, and it could be encouraging
+for users who come to PowerShell from bash since it is already familiar to
+them; however, the initial implementation of this operator was simply to
+invoke `Start-Job` in a manner familiar with bash users, and the current
+implementation doesn't work like the same operator does in bash. It also left
+several opportunities for a more feature-rich PowerShell-based version of that
+operator on the table. The end result is a feature that feels more like a
+checklist item that is partially checked off than a feature that adds
+significant, tangible value to the PowerShell community. In a nutshell, we can
+and should do better.
+
+Note that there is a [related RFC about adding a `~&` ThreadJob operator](https://github.com/PowerShell/PowerShell-RFC/pull/205).
+
+## Motivation
+
+As a user,<br/>
+I can reference the most recent job(s) launched with `Start-Job` or the `&` background operator in a `$!` variable<br/>
+So that I can write my handling of those jobs in a consistent way without the job object or id, using the same variable that is used in bash for the same purpose.
+
+As a user,<br/>
+I can see the output of jobs launched with the `&` background operator in my current host by default<br/>
+So that I can launch multiple background jobs and watch their concurrent progress in my host without blocking my console.
+
+As a user,<br/>
+I can launch jobs silently with a new `&!` background operator<br/>
+So that their output is only available when I use `Receive-Job` to receive that output.
+
+As a user,<br/>
+I can use the `&` operator (and the proposed `&!` operator) after the stop parsing sigil (`--%`)<br/>
+So that I can launch legacy commands in the background more easily, without having to use `Start-Job`.
+
+Also, for completeness/functional parity with cmdlets:
+
+As a user,<br/>
+I can see the output of jobs launched with `Start-Job` or `Start-ThreadJob` in my current host by using a new `-ShowInHost` switch<br/>
+So that I can launch multiple background jobs and watch their concurrent progress in my host without blocking my console.
+
+As a user,<br/>
+I can show or hide the output of running jobs by using new `Show-JobData` and `Hide-JobData` cmdlets<br/>
+So that I can control the display and monitoring of running jobs without having to muck around with `Receive-Job -Keep`.
+
+As a PowerShell contributor,<br/>
+I can see the process ID associated with any background job by looking at the job members<br/>
+So that I can more easily attach the Visual Studio debugger to that job when I need to.
+
+## User Experience
+
+### Launching a background job with the output visible in the host
+
+```powershell
+# This command:
+Get-Process &
+# Would return the same output as this command:
+Start-Job {Get-Process} -ShowInHost
```

Comment:
Since `Start-Job` defaults to show output, the `-ShowInHost` parameter is confusing. If I'm learning how to use `Start-Job`, it makes me think I need to use `-ShowInHost` all the time to get job data when I don't need to. Additionally, it would confuse me as to why it's giving me output when I didn't specify `-ShowInHost`

I suggest changing it to `-ShowProgress` or `-ShowStatus`. Progress has the benefit of describing the same behavior as `Write-Progress` making it more consistent with expectations.

Reply:
Agreed.

I just pushed several updates to the RFC to address this.

I flipped the recommendation around. It now leaves the `&` background operator behavior as is, which matches the default `Start-Job` behavior, and proposes `&!` be used when you want to see output (or the corresponding `Start-Job -ShowOutput` parameter.

I changed `-ShowInHost` to `ShowOutput`, and renamed `Show-/Hide-JobData` as `Show-/Hide-JobOutput` so that they are all consistent. I prefer `Output` as the suffix over `Progress` because this isn't the same as `Write-Progress`, and I don't want to confuse the two. I prefer `Output` as the suffix over `Status` because job status is a different thing.

I believe making this consistent on `Output` will set expectations appropriately and avoid confusion. If you disagree, please let me know.

Thank you for drawing my attention to this.